v4/export/dump.go

```
@@ -196,193 +188,386 @@ func (d *Dumper) Dump() (err error) {
+			return true
+		case irChan <- ir:
+			return false
+		default:
+			sleepTime := 1 * time.Second
+			time.Sleep(sleepTime)
+			log.Warn("write channel is full, waiting writer to consume IR",
+				zap.Int("channel size", defaultDumpThreads),
+				zap.Stringer("sleep time", sleepTime))
 		}
-		createTableSQL, err := ShowCreateTable(db, dbName, tableName)
+	}
+}
+
+var z = &big.Int{}
+
+func (d *Dumper) selectMinAndMaxIntValue(conn *sql.Conn, db, tbl, field string) (*big.Int, *big.Int, error) {
+	ctx, conf := d.ctx, d.conf
+	query := fmt.Sprintf("SELECT MIN(`%s`),MAX(`%s`) FROM `%s`.`%s`",
+		escapeString(field), escapeString(field), escapeString(db), escapeString(tbl))
+	if conf.Where != "" {
+		query = fmt.Sprintf("%s WHERE %s", query, conf.Where)
+	}
+	log.Debug("split chunks", zap.String("query", query))
+
+	var smin sql.NullString
+	var smax sql.NullString
+	row := conn.QueryRowContext(ctx, query)
+	err := row.Scan(&smin, &smax)
+	if err != nil {
+		log.Error("split chunks - get max min failed", zap.String("query", query), zap.Error(err))
+		return z, z, err
+	}
+	if !smax.Valid || !smin.Valid {
+		// found no data
+		log.Warn("no data to dump", zap.String("schema", db), zap.String("table", tbl))
+		return z, z, nil
+	}
+
+	max := new(big.Int)
+	min := new(big.Int)
+	var ok bool
+	if max, ok = max.SetString(smax.String, 10); !ok {
+		return z, z, errors.Errorf("fail to convert max value %s in query %s", smax.String, query)
+	}
+	if min, ok = min.SetString(smin.String, 10); !ok {
+		return z, z, errors.Errorf("fail to convert min value %s in query %s", smin.String, query)
+	}
+	return min, max, nil
+}
+
+func (d *Dumper) concurrentDumpTiDBTables(conn *sql.Conn, meta TableMeta, ir chan<- TableDataIR) error {
+	ctx, conf := d.ctx, d.conf
+	db, tbl := meta.DatabaseName(), meta.TableName()
+
+	handleColNames, handleVals, err := selectTiDBTableSample(db, tbl, conn, meta)
+	if err != nil {
+		return err
+	}
+	if len(handleVals) == 0 {
+		return nil
```

Should we return an error or some instructions here to help `sequentialDumpTable` or `concurrentDumpTable` using the unique index? This table will directly be skipped if we `return nil` here.

---

If the `handleVals` is empty, there is no data in this table.